### PR TITLE
AK+LibWeb: Move decode_forgiving_base64 to Web::Infra and fix remaining spec compliance problems

### LINKED ISSUE #16638: LibWeb: atob doesn't use the spec's "forgiving base64 decode" and other issues
For example:
```js
atob("YQ").length === 1; // should be 1, but we give 3.
atob("YR").length === 1; // should be 1, but we give 3.
atob("YQ") === atob("YR"); // should be true, but we give false.
```
Required by hCaptcha, which fails to JSON parse a string decoded by atob as it doesn't consume all the input:
```js
a = "eyJmIjowLCJzIjoyLCJ0IjoidyIsImQiOiJneDNpbkprMm16ZjVtN3JxMjlkK0oxSktETkRxZWVHZGN6YXlpVHJSQWpNbmxSNEwyd2lhU204OXFLaDF2SVdSRWNQY0IrV0NpOTRVVDZ3QWhsaGdzQ1ErdFpDeXNGVVZwa2JQZVludGUxdG5IQU9pUWtrODkxeGc5N0lxcXY4ZDFQZkpsa2FRa25FNUQvZzNkQ1VlMCt3ZCtEMzRMYTl3THZ0dzFESmkvY1k3TnJOLzJ2SEhyQkJuZnc9PVlpdEllbnpOeWZZWlZsalUiLCJsIjoiaHR0cHM6Ly9uZXdhc3NldHMuaGNhcHRjaGEuY29tL2MvMjAyN2Y4YyIsImUiOjE2NzE4MTA0NTYsIm4iOiJoc3ciLCJjIjoxMDAwfQ"
JSON.parse(atob(a)); // should parse correctly and give an object, but we throw with "Malformed JSON string"
```
We also don't throw InvalidCharacterError but instead TypeError on decoding failure.
We also decode the result with the windows-1252 decoder, which the spec doesn't do.

### PR DESCRIPTION
**Pull request by kuzux**
Instead of only stripping it from the ends, since that's actually what the spec says.

As noted in PR #16692 comments after it was merged; as well as issue #16638.

Fixes #16638

### CODE REVIEW

**Comment by kleinesfilmroellchen** [comment]
@kuzux please add a "Fixes #16638" so that the issue gets closed automatically

File: AK/Base64.cpp [hunk 1]
```diff
@@ -146,7 +146,13 @@ ErrorOr<String> encode_base64(ReadonlyBytes input)
 ErrorOr<ByteBuffer> decode_forgiving_base64(StringView input)
 {
     // 1. Remove all ASCII whitespace from data.
-    auto data = input.trim_whitespace();
+    // FIXME: It is possible to avoid copying input here, it's just a bit tricky to remove the equal signs
```
Comment (trflynn89) [inline AK/Base64.cpp:149]: I think these forgiving algorithms in general belong in LibWeb's Infra subfolder, something like LibWeb/Infra/Base64.cpp. For one, the spec's definition of whitespace is a bit different than `is_ascii_space` (See `Infra::is_ascii_whitespace`).

Comment (kuzux) [reply AK/Base64.cpp:149]: Well you're right (turns out the is ascii whitespace definitions are slightly different:D)

But I guess that could be a separate commit and PR

Comment (nico) [reply AK/Base64.cpp:149]: Maybe you could add the commit to this PR?

(I'll point out I said the same thing on the PR that added the function in the first place :P)

File: Userland/Libraries/LibWeb/Infra/Strings.cpp [hunk 2]
```diff
@@ -65,4 +66,129 @@ bool is_code_unit_prefix(StringView potential_prefix, StringView input)
...
+    for (size_t i = 0; i < base64_alphabet.size(); ++i) {
+        table[base64_alphabet[i]] = static_cast<i16>(i);
+    }
+    return table;
+}
+
+static constexpr auto base64_alphabet_lookup_table = make_base64_lookup_table();
+
+// https://infra.spec.whatwg.org/#forgiving-base64
+ErrorOr<ByteBuffer> decode_forgiving_base64(StringView input)
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/Infra/Strings.cpp:93]: Strings.cpp is section "4.6 Strings" of the Infra spec. Please put this in its own Base64.cpp file.

File: Userland/Libraries/LibWeb/HTML/Window.cpp [hunk 3]
```diff
@@ -1410,7 +1411,7 @@ JS_DEFINE_NATIVE_FUNCTION(Window::atob)
 
     // Otherwise, the user agent must convert data to a byte sequence whose nth byte is the eight-bit representation of the nth code point of data
     // and then must apply forgiving-base64 encode to that byte sequence and return the result.
-    auto decoded = AK::decode_forgiving_base64(StringView(deprecated_string));
+    auto decoded = Web::Infra::decode_forgiving_base64(StringView(deprecated_string));
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Window.cpp:1414]: No need to specify `Web::` within LibWeb

File: Userland/Libraries/LibWeb/Infra/Base64.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+{
+    Array<i16, 256> table;
+    table.fill(-1);
+    for (size_t i = 0; i < base64_alphabet.size(); ++i) {
+        table[base64_alphabet[i]] = static_cast<i16>(i);
+    }
+    return table;
+}
+
+static constexpr auto base64_alphabet_lookup_table = make_base64_lookup_table();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Infra/Base64.cpp:39]: can we keep this in AK and just export it for global use?

File: Userland/Libraries/LibWeb/Infra/Base64.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+// https://infra.spec.whatwg.org/#forgiving-base64
+ErrorOr<ByteBuffer> decode_forgiving_base64(StringView input)
+{
+    // 1. Remove all ASCII whitespace from data.
+    // FIXME: It is possible to avoid copying input here, it's just a bit tricky to remove the equal signs
+    StringBuilder builder;
+    for (auto character : input) {
+        if (!is_ascii_space(character))
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Infra/Base64.cpp:48]: Please put the commit moving it first and then use the proper definition of ASCII whitespace in the fix

File: Userland/Libraries/LibWeb/Infra/Base64.h [hunk 6]
```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright (c) 2022-2023, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Forward.h>
+
+namespace Web::Infra {
+ErrorOr<ByteBuffer> decode_forgiving_base64(StringView input);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Infra/Base64.h:10]: Missing newlines around namespace curly braces

File: Userland/Libraries/LibWeb/Infra/Strings.cpp [hunk 7]
```diff
@@ -5,6 +5,7 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <AK/CharacterTypes.h>
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Infra/Strings.cpp:8]: what's this for?

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/Infra/Strings.cpp (hunk 7 above)
Comment (kuzux) [reply Userland/Libraries/LibWeb/Infra/Strings.cpp:8]: Just an artifact from moving the original function a couple of times, will remove :)

File: Userland/Libraries/LibWeb/Infra/Base64.h [hunk 8]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2022-2023, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Infra/Base64.h:6]: Linter is complaining about missing `#pragma once` here

**Review by linusg — COMMENTED** [review]
Thanks, that's much better :)

File: Userland/Libraries/LibWeb/Infra/Base64.h [hunk 9]
```diff
@@ -0,0 +1,16 @@
+#
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Infra/Base64.h:1]: `Files with bad licenses: Userland/Libraries/LibWeb/Infra/Base64.h`

**Pull request merged**